ArchiSteamFarm.Tests/Bot.cs

```diff
@@ -0,0 +1,386 @@
+		}
+
+		[TestMethod]
+		public void OtherAppIDOneSet() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, realAppID: 42),
+				CreateCard(2, realAppID: 42),
+				CreateCard(1, realAppID: 43),
+				CreateCard(2, realAppID: 43),
+				CreateCard(3, realAppID: 43),
+				CreateCard(1, realAppID: 44),
+				CreateCard(2, realAppID: 44),
+				CreateCard(3, realAppID: 44),
+				CreateCard(4, realAppID: 44)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(
+				items, new Dictionary<uint, byte> {
+					{ 42, 3 },
+					{ 43, 3 },
+					{ 44, 3 }
+				}
+			);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (43, Steam.Asset.SteamCommunityContextID, 1), 1 },
+				{ (43, Steam.Asset.SteamCommunityContextID, 2), 1 },
+				{ (43, Steam.Asset.SteamCommunityContextID, 3), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		[TestMethod]
+		public void TooHighAmount() {
+			HashSet<Steam.Asset> items = new HashSet<Steam.Asset> {
+				CreateCard(1, 2),
+				CreateCard(2)
+			};
+
+			HashSet<Steam.Asset> itemsToSend = GetItemsForFullBadge(items, cardsPerSet: 2);
+
+			Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult = new Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> {
+				{ (42, Steam.Asset.SteamCommunityContextID, 1), 1 },
+				{ (42, Steam.Asset.SteamCommunityContextID, 2), 1 }
+			};
+
+			AssertResultMatchesExpectation(expectedResult, itemsToSend);
+		}
+
+		private static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, uint appID = (uint) 42, byte cardsPerSet = byte.MaxValue) => GetItemsForFullBadge(inventory, new Dictionary<uint, byte> { { appID, cardsPerSet } });
+
+		private static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, IDictionary<uint, byte> cardsPerSet) {
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), List<uint>> inventorySets = ArchiSteamFarm.Trading.GetInventorySets(inventory);
+
+			return ArchiSteamFarm.Bot.GetItemsForFullBadge(inventory, inventorySets.ToDictionary(kv => kv.Key, kv => (SetsToExtract: inventorySets[kv.Key][0], cardsPerSet[kv.Key.RealAppID]))).ToHashSet();
+		}
+
+		private static Steam.Asset CreateCard(ulong classID, uint amount = 1, uint realAppID = (uint) 42, Steam.Asset.EType type = Steam.Asset.EType.TradingCard, Steam.Asset.ERarity rarity = Steam.Asset.ERarity.Common) => new Steam.Asset(Steam.Asset.SteamAppID, Steam.Asset.SteamCommunityContextID, classID, amount, realAppID: realAppID, type: type, rarity: rarity);
+
+		private static void AssertResultMatchesExpectation(Dictionary<(uint RealAppID, ulong ContextID, ulong ClassID), uint> expectedResult, IEnumerable<Steam.Asset> itemsToSend) {
```

Comment:
Change that input to `IReadOnlyDictionary` and `IReadOnlyCollection` if possible, throw on nulls, allow empty collections.